article: add "A clock based on Cistercian numerals" (#26)
hsablonniere/hsablonniere.com: pages/a-clock-based-on-cistercian-numerals--hptit8/index.md

---
id: hptit8
title: A clock based on cistercian numerals
description: TODO
date: 2021-03-17
tags: article
layout: article.njk
withCodeSamples: true
stylesheets:
- ./index.css
scripts:
- ./node_modules/cistercian-numerals/dist/cistercian-numerals.js
---

> Let me tell you the story of when I tried Twitch for the first time to live-code a Web Component that diplays numbers with cistercian numerals, so I could build a clock like this:

Reviewer: ```suggestion
Let me tell you the story of when I tried Twitch for the first time to live-code a Web Component that displays numbers with Cistercian numerals, so I could build a clock like this:
```

<cistercian-clock></cistercian-clock>

## Yet another side project idea

> Back in january 2021, I saw [a tweet](https://twitter.com/Rainmaker1973/status/1348307518045507584) about cistercian numerals.

Reviewer: ```suggestion
Back in January 2021, I saw [a tweet](https://twitter.com/Rainmaker1973/status/1348307518045507584) about Cistercian numerals.
```

I was very intrigued.
If you don't know about them, here's the introduction from the [Wikipedia page](https://en.wikipedia.org/wiki/Cistercian_numerals):

> The medieval Cistercian numerals, or "ciphers" in nineteenth-century parlance, were developed by the Cistercian monastic order in the early thirteenth century at about the time that Arabic numerals were introduced to northwestern Europe. They are more compact than Arabic or Roman numerals, with a single character able to indicate any integer from 1 to 9,999.

They look like this:

![Some numbers represented with cistercian numerals: 1, 20, 400, 4000, 5555, 6789, 9394](./wikipedia-cistercian-numerals.png)

I really like how compact and elegant it is to display a big number with just one symbol, especially when there is some symmetry.

> A few days later, I saw [a tweet](https://twitter.com/aqandrew/status/1349762018639638528) from [Andrew Aquino](https://twitter.com/aqandrew) saying he created a React component to display any number with cistercian numerals.

Reviewer: ```suggestion
A few days later, I saw [a tweet](https://twitter.com/aqandrew/status/1349762018639638528) from [Andrew Aquino](https://twitter.com/aqandrew) saying he created a React component to display any number with Cistercian numerals.
```

My [first reaction](https://twitter.com/hsablonniere/status/1357079582499876867) was to tweet:

> sudo make web-component

> I was expecting someone to create a Web Component, so we could all have a standard compatible way to display cistercian numbers on our websites, even if we don't use React.

Reviewer: ```suggestion
I was expecting someone to create a Web Component, so we could all have a standard compatible way to display Cistercian numbers on our websites, even if we don't use React.
```

Instead, out of distaste for `sudo`, my wonderful colleague [Marc-Antoine Perennou](https://twitter.com/Keruspe/) created [sudo-ng](https://github.com/Keruspe/sudo-ng).

> Nothing happened, so I guess I was left with yet another side project idea, but this time, I decided to do it live with an audience.

Reviewer: ```suggestion
Nothing happened, so I guess I was left with yet another side project idea.
Although this time, I decided to do it live with an audience.
```

## Trying Twitch for the first time

In the pre-pandemic world, I used to give talks at conferences and share my knowledge and passion for Web development.

> When you code in from of hundreds of people in a conference room, everything is well-prepared and rehearsed many times.

Reviewer: ```suggestion
When you code in front of hundreds of people in a conference room, everything is well-prepared and rehearsed many times.
```

> There is a kind of illusion of simplicity, but it's the best way to go straight to the point and focus on the new stuffs the audience wants to learn about.

Reviewer: ```suggestion
There is a kind of illusion of simplicity, but it's the best way to go straight to the point and focus on the new stuff the audience wants to learn about.
```

> The minor drawback is that people miss everything that happen in real life when you write code.

Reviewer: ```suggestion
The minor drawback is that people miss everything that happens in real life when you write code.
```

You read some docs, you make mistakes, you try something, you go back, you get stuck on a silly bug...

Watching someone thinking out loud and writing code without preparation is a wonderful way to learn.
You discover new IDE keyboard shortcuts.
You learn about some unknown language pattern or feature.
It feels a bit like pair-programming with a mentor.

Following a few [fellow](https://www.twitch.tv/nullpointeur) [speakers](https://www.twitch.tv/k33g_org) that got into Twitch during the first lockdown, I decided to create [my channel](https://www.twitch.tv/hsablonniere) and try the platform for this side project idea.

My main goal was to create a Web Component to display a number with cistercian numerals and then reuse it to create a clock component that displays three cistercian numbers: hours, minutes and seconds.
Bonus points if I could animate the lines of the symbols.

## The live-coding session

If you're interested, I posted a [recording of the session](https://www.youtube.com/watch?v=88AKamcJ-bY) on YouTube but beware:

> * 👂 It's in french.

Reviewer: ```suggestion
* 👂 It's in French.
```

> * 👂 It's in french.

Reviewer: ```suggestion
* 👂 It's in French.
```

* ⌛ It's 2 hours long (<cistercian-number inline value="120">120</cistercian-number> minutes)
* 💬 The chat is NOT displayed.
* 😱 I spent way too much time on Inkscape trying to generate the initial SVG design.
* 🚧 I should have built both components before thinking about animating them.

Despite those, I really enjoyed giving this session.
Mistakes were made, but we finished with some kind of achievement.
The clock design was a bit clunky, but it worked.
It looked like this:

<mark>TODO: VIDEO OF THE PROJECT AT THE END OF THE STREAM</mark>

If I were to do it again, I would prioritize the different steps, so the progress is more linear for the audience.

> I would also improve the way I time box steps, so I would stop insisting on some details.

Reviewer: ```suggestion
I would also improve the way I timebox steps, so I would stop insisting on some details.
```

I think it's just a different way of sharing knowledge with others.
A bit slower, more interactive, less magic but definitely more authentic.

## Post-stream improvements

After the stream, I decided to improve the project:

* Implement the numbers after <cistercian-number inline value="99">99</cistercian-number>
* Fix the positioning of the different lines
* Add some animations using the [animated line drawing SVG](https://jakearchibald.com/2013/animated-line-drawing-svg/) technique 

A few weeks after, I added a few more features:

* Use CSS custom properties to allow color customization of each lines 
* Use CSS custom properties to allow line width customization 
* Add `date` mode for the clock to display year, month, day before hours, minutes, seconds
* Add `no-seconds` mode for the clock to remove the seconds
* Add `inline` mode for the number, so you can use a number inside a paragraph of text (as featured in this article)
* Try to improve the accessibility (any feedbacks on this is welcome)

I published the source on GitHub at [hsablonniere/cistercian-numerals](https://github.com/hsablonniere/cistercian-numerals) if you're curious.

## Web Components are awesome!

> Because they're exposed as Web components, you can use them in any Web project, whatever the stack/framework you're using.

Reviewer: ```suggestion
Because they're exposed as Web Components, you can use them in any Web project, whatever the stack/framework you're using.
```

You can install them [from npm](https://www.npmjs.com/package/cistercian-numerals) but it's not mandatory at all.

> As described in the [README](https://github.com/hsablonniere/cistercian-numerals/blob/master/README.md), you can use those components in your HTML without any `npm install`, without any build step or bundler.

Reviewer: ```suggestion
As described in the [README](https://github.com/hsablonniere/cistercian-numerals/blob/master/README.md), you can use these components in your HTML without any `npm install`, without any build step or bundler.
```
